### PR TITLE
Chapter2 solutions added

### PR DESCRIPTION
**Pull request by wookenny**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +347,10 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList start end list = if start < 0 then [] else take len (drop start list)
+    where
+      len = end - start + 1
```
Comment (chshersh) [inline src/Chapter2.hs:352]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,8 +637,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
-
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x:_:xs) = x : takeEven xs
```
Comment (chshersh) [inline src/Chapter2.hs:643]: Pattern-matching award! :trophy:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -727,7 +746,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\el -> replicate el el) l)
```
Comment (chshersh) [inline src/Chapter2.hs:749]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 4]
```diff
@@ -740,7 +759,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains el l = filter (elem el) l
```
Comment (chshersh) [inline src/Chapter2.hs:763]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 5]
```diff
@@ -780,13 +800,13 @@ Let's now try to eta-reduce some of the functions and ensure that we
...
+divideTenBy = div 10
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Ord a => a -> [a] -> [a]
+listElementsLessThan x = filter (< x)
 
--- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: Num a => [a] -> [a] -> [a]
```
Comment (chshersh) [inline src/Chapter2.hs:808]: Nice work on the type signature here!

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,7 +881,11 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind l =  x : rewind xs
+    where 
+      (x:xs) = rotate (length l -1) l
```
Comment (chshersh) [inline src/Chapter2.hs:888]: I'm afraid, this implementation will be slow. `rotate` works in linear time depending on the list size, rotating the list on each recursive call is very slow...

**Review by chshersh — APPROVED** [review]
Nice work on solving challenges for Chapter Two!
And well done with the Advanced Task :ok: 

I left my review and some comments. Hope you find them helpful :slightly_smiling_face:

File: src/Chapter2.hs (hunk 4 above)
Comment (wookenny) [reply src/Chapter2.hs:763]: So easy to miss with my limited experience, thanks for pointing out.

File: src/Chapter2.hs (hunk 6 above)
Comment (wookenny) [reply src/Chapter2.hs:888]: No I get how to `rewind` in linear time, thanks for pointing out.

**Comment by wookenny** [comment]
Thanks for the review, helped me out a lot. :cake:

**Pull request merged**